File: java/com/epicdragonworld/gameserver/managers/ChatManager.java
package com.epicdragonworld.gameserver.managers;

import com.epicdragonworld.gameserver.model.Location;
import com.epicdragonworld.gameserver.model.actor.Player;
import com.epicdragonworld.gameserver.network.packets.sendable.ChatResult;

/**
 * @author Pantelis Andrianakis
 */
public class ChatManager
{
	private static final byte CHAT_TYPE_SYSTEM = 0;
	private static final byte CHAT_TYPE_NORMAL = 1;
	private static final byte CHAT_TYPE_MESSAGE = 2;
	private static final String COMMAND_PERSONAL_MESSAGE = "/tell ";
	private static final String COMMAND_LOCATION = "/loc";
	private static final String SYS_NAME = "System";
	private static final String MSG_TO = "To ";
	
	public static void handleChat(Player sender, String message)
	{
		// Check if message is empty.
		message = message.trim();
		if (message.isEmpty())
		{
			return;
		}
		
		final String lowercaseMessage = message.toLowerCase().replaceAll("\\s{2,}", " "); // Also remove all double spaces.
		if (lowercaseMessage.equals(COMMAND_LOCATION))
		{
			final Location location = sender.getLocation();
			sender.channelSend(new ChatResult(CHAT_TYPE_SYSTEM, SYS_NAME, "Your location is " + location.getX() + " " + location.getY() + " " + location.getZ()));
		}
		else if (lowercaseMessage.startsWith(COMMAND_PERSONAL_MESSAGE))
		{
			final String[] lowercaseMessageSplit = lowercaseMessage.split(" ");
			if (lowercaseMessageSplit.length < 3) // Check for parameters.
			{
				sender.channelSend(new ChatResult(CHAT_TYPE_SYSTEM, SYS_NAME, "Incorrect syntax. Use /tell [name] [message]."));
				return;
			}
			
			final Player receiver = WorldManager.getPlayerByName(lowercaseMessageSplit[1]);
			if (receiver == null)
			{
				sender.channelSend(new ChatResult(CHAT_TYPE_SYSTEM, SYS_NAME, "Player was not found."));
			}
			else
			{
				// Step by step cleanup, to avoid problems with extra/double spaces on original message.
				message = message.substring(lowercaseMessageSplit[0].length(), message.length()).trim(); // Remove command.
				message = message.substring(lowercaseMessageSplit[1].length(), message.length()).trim(); // Remove receiver name.
				sender.channelSend(new ChatResult(CHAT_TYPE_MESSAGE, MSG_TO + receiver.getName(), message));
				receiver.channelSend(new ChatResult(CHAT_TYPE_MESSAGE, sender.getName(), message));
			}
		}
		else // Normal message.
		{
			sender.channelSend(new ChatResult(CHAT_TYPE_NORMAL, sender.getName(), message));
			for (Player player : WorldManager.getVisiblePlayers(sender))
			{
				player.channelSend(new ChatResult(CHAT_TYPE_NORMAL, sender.getName(), message));
			}
		}
	}
}

File: EpicDragonWorldServer/Managers/ChatManager.cs
/**
 * Author: Pantelis Andrianakis
 * Date: November 7th 2018
 */
public class ChatManager
{
    private static readonly byte CHAT_TYPE_SYSTEM = 0;
    private static readonly byte CHAT_TYPE_NORMAL = 1;
    private static readonly byte CHAT_TYPE_MESSAGE = 2;
    private static readonly string SYS_NAME = "System";
    private static readonly string MSG_TO = "To ";
    // Normal player commands
    private static readonly string COMMAND_PERSONAL_MESSAGE = "/tell ";
    private static readonly string COMMAND_LOCATION = "/loc";
    private static readonly string COMMAND_RETURN = "/return";
    // Administrator commands
    private static readonly string COMMAND_SPAWN = "/spawn ";

    public static void HandleChat(Player sender, string message)
    {
        // Check if message is empty.
        message = message.Trim();
        if (message.Length == 0)
        {
            return;
        }

        bool isAdmin = sender.GetAccessLevel() > 99;
        string lowercaseMessage = message.ToLowerInvariant().Replace("\\s{2,}", " "); // Also remove all double spaces.
        if (lowercaseMessage.Equals(COMMAND_LOCATION))
        {
            LocCommand.Handle(sender);
        }
        else if (lowercaseMessage.Equals(COMMAND_RETURN))
        {
            ReturnCommand.Handle(sender);
        }
        else if (lowercaseMessage.StartsWith(COMMAND_PERSONAL_MESSAGE))
        {
            TellCommand.Handle(sender, lowercaseMessage, message);
        }
        else if (isAdmin && lowercaseMessage.StartsWith(COMMAND_SPAWN))
        {
            SpawnCommand.Handle(sender, lowercaseMessage);
        }
        else // Normal message.
        {
            sender.ChannelSend(new ChatResult(CHAT_TYPE_NORMAL, sender.GetName(), message));
            foreach (Player player in WorldManager.GetVisiblePlayers(sender))
            {
                player.ChannelSend(new ChatResult(CHAT_TYPE_NORMAL, sender.GetName(), message));
            }
            // Log chat.
            if (Config.LOG_CHAT)
            {
                LogManager.LogChat("[" + sender.GetName() + "] " + message);
            }
        }
    }

    public static void SendPrivateMessage(Player sender, Player receiver, string message)
    {
        sender.ChannelSend(new ChatResult(CHAT_TYPE_MESSAGE, MSG_TO + receiver.GetName(), message));
        receiver.ChannelSend(new ChatResult(CHAT_TYPE_MESSAGE, sender.GetName(), message));
    }

    public static void SendSystemMessage(Player player, string message)
    {
        player.ChannelSend(new ChatResult(CHAT_TYPE_SYSTEM, SYS_NAME, message));
    }
}
